### ISSUE TITLE
Use Bit-Vector for Initialized Flags 

### ISSUE THREAD

**Issue by ibraheemdev** [ibraheemdev/boxcar#7]
Right now each entry has an `AtomicBool`, which can waste a lot of memory depending on the alignment of `T`.

**Comment by clarfonthey** [comment]
I agree that would be nice, although it'd require  although it would require using unstable APIs to make it efficient. You'd have to leverage the existing pointer to prefix the entry array with a bit-vector of sufficient length, and there's really no easy way to make this work on stable.

Additionally, I'm not sure how this would affect cache locality, since each entry's flag is no longer right next to it, and larger buckets could exceed the size to fit an entry's flag and its value in cache.

**Comment by ibraheemdev** [comment]
> You'd have to leverage the existing pointer to prefix the entry array with a bit-vector of sufficient length, and there's really no easy way to make this work on stable.

You can do with using `alloc` directly with a custom layout.

> Additionally, I'm not sure how this would affect cache locality, since each entry's flag is no longer right next to it, and larger buckets could exceed the size to fit an entry's flag and its value in cache.

It's true that the flag and value would no longer be in the same cache line. One thing we could do to improve this is to have a true marker length that's lazily updated based on updates to the bitvector, such that everything up to the marker length is initialized. This would reduce contention for readers.

> Also, if you're using this crate, you should know that it's not memory-efficient. You sacrifice memory efficiency and the ability to keep everything in one place with concurrency.

Using a bitvec can cut the memory overhead almost in half in many cases, so I think it's worth doing.

**Comment by clarfonthey** [comment]
Oh huh, I guess I never realised that even though `Allocator` isn't stable, `alloc` is, so, never mind. You definitely could do this with a custom layout, although you'd unfortunately have to represent said layout using a struct like:

```rust
#[repr(C)]
struct Entry<T, const N: usize> {
    active: BitArr![N, in AtomicU8],
    data: T,
}
```

since I don't think we can really guarantee that an unsized slice of data would always be at the end.
